lib/ukboot/boot.c

```diff
@@ -303,6 +303,7 @@ void ukplat_entry(int argc, char *argv[])
 	if (!tls)
 		UK_CRASH("Failed to allocate and initialize TLS\n");
 
+	memset(tls, 0, ukarch_tls_area_size());
```

Comment:
@skuenzer  Hi. Thank you for response. The original issue appeared on arm64, but I don't think it is arch specific. Per my understanding, the code snippets you are referring to, they clear tbss, what is per thread bss section (uninitialized global data with __thread specifier goes there). However, I'm talking about pthread_t structure which resides in tls and used by pthreads. So these are different part of tls memory. Maybe I need to add a little bit more context how is that reproduced with pthreads.

Reply:
Okay :thinking:, this should be done by musl that should provide `ukarch_tls_tcb_init()` (expected by [arch/x86/x86_64/tls.c:201](https://github.com/unikraft/unikraft/blob/staging/arch/x86/x86_64/tls.c#L201)): https://github.com/unikraft/lib-musl/blob/staging/__uk_init_tls.c#L177 . Can you check this function?

Reply:
@skuenzer maybe you mean ukarch_tls_tcb_init -> __uk_init_tls -> __uk_init_tp ? I don't see how uk_thread_uktcb_init (https://github.com/unikraft/lib-musl/blob/staging/__uk_init_tls.c#L177) appears in this chain. In this 3 functions I have mentioned above I don't see any tls cleaning.

Reply:
I think the problem is that the pthread structure is not initialized in the main thread. So zeroing the structure can fix some of the problems but in the end there are also list structures, which must not be initialized to zero. I think we could call `uk_thread_uktcb_init` for the main thread.

Reply:
@mschlumpp yes the problem is in non initialized `pthread_t` for main thread. Okay, then let me look how exactly `uk_thread_uktcb_init` can be used for init thread here.

Reply:
@alexhoppus Sorry, my mistake. You are right, it is [lib-musl:__uk_init_tls.c:L219](https://github.com/unikraft/lib-musl/blob/staging/__uk_init_tls.c#L219) that then calls `__uk_init_tls()` and `__uk_init_tp()`. In fact, some clearing is missing here, so you should patch somewhere here.
Normally, `ukarch_tls_tcb_init()` is called for every thread before `uk_thread_uktcb_init()` is called for thread creations using `lib/uksched`-API calls. pthreads initialize the TCB manually behind musl's pthread APIs but the main thread only goes through `ukarch_tls_tcb_init()`. So, it makes sense to initialize a zeroed (or somehow meaningful initialized) pthread structure for fixing the problem.
Ideally, such a fix may also be an enabler for builds with musl but without threading (`lib/uksched`). As correct behavior, the pthread API could then not be used, the calls should fail because thread creation is not supported in such a case. As background: We ignored musl without `lib/uksched` for the first version but we intended to enable it later again.

Reply:
@skuenzer I think we should initialize pthread in `__uk_init_tp()` in the same way as pthread_create do. What do you think?

Reply:
I have opened new PR https://github.com/unikraft/lib-musl/pull/59 in musl

Reply:
Yes, I agree, great. This means, we can close this PR and continue review with [PR #59](https://github.com/unikraft/lib-musl/pull/59) with lib-musl.